docs/config_system.md

```diff
@@ -0,0 +1,270 @@
+- `target_overrides` is a dictionary with target-specific values for the configuration parameters.
+
+`target_overrides` is used to override the values of the parameters depending on the current compilation target. The keys in `target_overrides` are matched against toolchain *labels* (a description of mbed targets can be found [here](mbed_targets.md)). If a key inside `target_overrides` matches one of the target labels, the parameter values are changed according to the value of the key. In the example above:
+
+- `config` is always processed first, independent of the target. `config` might define values for some of the parameters. In this case, `buffer_size` will be set to 1024, `queue_size` will be set to 10 and `timer_period` will not have a value.
+- if the library is compiled for the `K64F` target, `timer_period` will be set to 100 and `queue_size` will be set to 40, since they are overridden by the `K64F` key in `target_overrides`. `buffer_size` will be set to 1024, as defined in `config`.
+- assuming that `NXP` is a label defined by **all** NXP based targets, if the library is compiled for **any** `NXP` target (like `LPC1768` or `LPC11U24`), `buffer_size` will be set to 128 and `queue_size` will be set to 20, while `timer_period` will not have a value (since it doesn't get one neither in `config`, nor in the `NXP` override).
+- the keys in `target_overrides` are processed in order: if a hypothetical target defines both `K64F` and `NXP` as labels, `timer_period` will be set to 100, `queue_size` will be set to 20 and `buffer_size` will be set to 128.
+- if the library is compiled for a target that doesn't have `K64F` or `NXP` as labels, the values of the parameters will be the ones set in `config`.
+
+Except `name`, all the above keys in the JSON file are optional, but if `target_overrides` is defined, `config` must also be defined.
+
+As explained [here](#defining-configuration-parameters), the parameters have an implicit `mylib.` prefix. Outside `mylib`, `buffer_size` is accessible using the name `mylib.buffer_size`. An application will be able to override the value of this parameter, as described in [this section](#configuration-data-in-applications).
+
+ If the source tree has code for more than one library, each library needs its own `mbed_lib.json` file in its root folder.
+
+# Configuration data in targets
+
+Like libraries, targets can define their own configuration data. Additionally, tables can override the configuration of the target(s) they inherit from (for more details about how do define a target and target inheritance, check [this link](mbed_targets.md)). Target configuration data is defined in `targets.json` using `config`, as described [here](#defining-configuration-parameters). An example for a hypothetical `Base` target is given below:
+
+```
+"Base": {
```

Comment:
> If this is the case then we need to turn ALL targets in targets.json into configuration data and I don't think that this is sensible or needed.

We definitely don't need to do that. Targets have their own legacy configuration, which is there before the configuration system existed. The configuration system doesn't try to change that, it just adds the mechanisms to add more specific configuration **in a structured way** (which means that there are clear rules for overriding configuration data up to the application layer). There is some overlap between configuation data and the previous mechanism, as you say, but it doesn't mean in any way that we **need** to convert them.

Also, if you want to -1 something, it makes much more sense to -1 the implementation, not the associated docs (especially since this was already reviewed and approved elsewhere).